cli/src/commands/publish.js

```diff
@@ -0,0 +1,269 @@
+}
+
+const validateFields = (
+    config,
+    fieldNames = ['id', 'version', 'minDHIS2Version']
+) => {
+    fieldNames.forEach(fieldName => {
+        const fieldValue = config[fieldName]
+        if (
+            (requiredFields.has(fieldName) && fieldValue == null) ||
+            fieldValue === ''
+        ) {
+            reporter.error(
+                `${fieldName} not found in config. Add an ${chalk.bold(
+                    fieldName
+                )}-field to ${chalk.bold('d2.config.js')}`
+            )
+            process.exit(1)
+        }
+        const fieldValidation = configFieldValidations[fieldName]
+        if (fieldValidation) {
+            const valid = fieldValidation(fieldValue)
+            if (typeof valid === 'string') {
+                reporter.error(`${fieldName}: ${valid}`)
+                process.exit(1)
+            } else if (!valid) {
+                reporter.error(`Invalid ${fieldName}`)
+                process.exit(1)
+            }
+        }
+    })
+}
+
+const resolveBundle = (cwd, params) => {
+    const appBundle = {}
+    // use file-path from params
+    if (params.file) {
+        try {
+            const filePath = path.resolve(cwd, params.file)
+            if (!fs.statSync(filePath).isFile()) {
+                reporter.error(`${params.file} is not a file`)
+                process.exit(1)
+            }
+            appBundle.id = params.appId
+            appBundle.version = params.fileVersion
+            appBundle.path = filePath
+            appBundle.name = path.basename(filePath)
+            appBundle.minDHIS2Version = params.minDHIS2Version
+            appBundle.maxDHIS2Version = params.maxDHIS2Version
+        } catch (e) {
+            reporter.error(`File does not exist at ${params.file}`)
+            process.exit(1)
+        }
+    } else {
+        // resolve file from built-bundle
+        const paths = makePaths(cwd)
+        const builtAppConfig = parseConfig(paths)
+        validateFields(builtAppConfig)
+
+        appBundle.id = builtAppConfig.id
+        appBundle.version = builtAppConfig.version
+        appBundle.path = path.relative(
+            cwd,
+            paths.buildAppBundle
+                .replace(/{{name}}/, builtAppConfig.name)
+                .replace(/{{version}}/, builtAppConfig.version)
+        )
+        appBundle.name = builtAppConfig.name
+        appBundle.minDHIS2Version = builtAppConfig.minDHIS2Version
+        appBundle.maxDHIS2Version = builtAppConfig.maxDHIS2Version
+
+        if (!fs.existsSync(appBundle.path)) {
+            reporter.error(
+                `App bundle does not exist, run ${chalk.bold(
+                    'd2-app-scripts build'
+                )} before deploying.`
+            )
+            process.exit(1)
+        }
+    }
+
+    return appBundle
+}
+
+const promptForConfig = async params => {
+    if (process.env.CI && (!params.apikey || !params.minVersion)) {
```

Comment:
> @Birkbjo I think this will throw if the API key is passed as an env var in CI, right? We actually should prefer to use env vars in CI

Agree that we should prefer env-vars. But good catch, this would work with the previous env-var, as that would be resolved by yargs - but I totally forgot that this would not work when I changed the name.